### PR TITLE
Add method of statically linking discord-rpc and QtApng

### PR DESCRIPTION
**Pull request by skyedeving**
Add submodules to build the libraries statically and a DEFINE option to link to them. No need to bundle these libs anymore
# Steps
```sh
git submodule update --init --recursive
```
## QtApng
```sh
patch external/QtApng/.qmake.conf external/static-QtApng.patch
cd external/QtApng
qmake
make
```

## discord-rpc
```sh
cd external/discord-rpc && mkdir build && cd build
cmake ..
make
```

## AO2-Client
Uncomment `DEFINE += STATIC_LIB` in `Attorney_Online.pro`
```sh
qmake
make
```

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -26,8 +26,12 @@ int main(int argc, char *argv[])
   QSettings *configini = main_app.configini;
 
   QPluginLoader apngPlugin("qapng");
+#ifdef STATIC_LIB
+  Q_IMPORT_PLUGIN(ApngImagePlugin)
```
Comment (oldmud0) [inline src/main.cpp:30]: From my understanding, this goes immediately after the includes, not inside `main()`.

File: Attorney_Online.pro [hunk 2]
```diff
@@ -26,6 +26,28 @@ contains(DEFINES, DISCORD) {
...
+# cmake ..
+# make
+#
+# DEFINES += STATIC_LIB
+contains(DEFINES, STATIC_LIB) {
+  LIBS += -L$$PWD/external/QtApng/plugins/imageformats -L$$PWD/external/QtApng/lib
+  contains(DEFINES, DISCORD) {
+    LIBS += -L$$PWD/external/discord-rpc/build/src
+  }
+  LIBS += -l:libqapng.a -l:libqtapng_png.a -l:libqtapng_zlib.a
```
Comment (oldmud0) [inline Attorney_Online.pro:48]: Oh, I remember why I did not do static links of QtApng on Windows builds. There is a nasty issue where the entry point of the libpng self-test utility gets linked into the program and set as the main entry point of the application. Please check that this doesn't happen on Windows.

Comment (skyedeving) [reply Attorney_Online.pro:48]: I'm going to need help with this from someone. I don't have a windows dev environment setup for this.

**Comment by skyedeving** [comment]
Accidentally deleted this branch

**Comment by skyedeving** [comment]
I tried to build Attorney_Online with statically built Qt5 but gave up. Here are the static builds of Qt5 created through github actions although:

https://github.com/skyedeving/qt5/releases/tag/5.15.2-github
https://github.com/skyedeving/qt5/blob/5.15.2-github/.github/workflows/static-qt.yml

**Comment by oldmud0** [comment]
> 
> I tried to build Attorney_Online with static linking to Qt5 but gave up. I managed to get one for linux statically linked to Qt5 but it required yet another static link to a specific version of a lib called libicuuc. Incredibly annoying.
>

I don't recommend making static builds of AO anymore, especially on Linux. I recommend investigating the RPATH trick that was suggested to me by in1tiate, and then using windeployqt/linuxdeployqt to copy the right DLLs to some subfolder.

**Comment by skyedeving** [comment]
Yeah I spent a lot of time getting one build of the client statically linked to qt and it didn't even match my distro's theme. Builds of qt5 static do exist though for anyone who wants to try linking to them in the future. The final step would've been linking a static build of libicu66-1 I think since the client was working when linking to that dynamically.

**Pull request closed without merging**